funcbench/bench.go

```diff
@@ -0,0 +1,142 @@
+// Copyright 2020 The Prometheus Authors
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+package main
+
+import (
+	"bytes"
+	"encoding/base64"
+	"fmt"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+	"strings"
+	"time"
+
+	"github.com/pkg/errors"
+	"golang.org/x/tools/benchmark/parse"
+	"gopkg.in/src-d/go-git.v4"
+	"gopkg.in/src-d/go-git.v4/plumbing"
+)
+
+// TODO: Add unit test.
+type Benchmarker struct {
+	logger Logger
+
+	benchFunc      string
+	benchTime      time.Duration
+	benchTimeout   time.Duration
+	resultCacheDir string
+
+	c    *commander
+	repo *git.Repository
+}
+
+func newBenchmarker(logger Logger, env Environment, c *commander, benchTime time.Duration, benchTimeout time.Duration, resultCacheDir string) *Benchmarker {
+	return &Benchmarker{
+		logger:         logger,
+		benchFunc:      env.BenchFunc(),
+		benchTime:      benchTime,
+		benchTimeout:   benchTimeout,
+		c:              c,
+		repo:           env.Repo(),
+		resultCacheDir: resultCacheDir,
+	}
+}
+
+func (b *Benchmarker) benchOutFileName(commit plumbing.Hash) (string, error) {
+	// Sanitize bench func.
+	bb := bytes.Buffer{}
+	e := base64.NewEncoder(base64.StdEncoding, &bb)
+	if _, err := e.Write([]byte(b.benchFunc)); err != nil {
+		return "", err
+	}
+	if err := e.Close(); err != nil {
+		return "", err
+	}
+
+	return fmt.Sprintf("%s-%s.out", bb.String(), commit.String()), nil
+}
+
+func (b *Benchmarker) execBenchmark(pkgRoot string, commit plumbing.Hash) (out string, err error) {
+	fileName, err := b.benchOutFileName(commit)
+	if err != nil {
+		return "", err
+	}
+
+	if _, err := ioutil.ReadFile(filepath.Join(b.resultCacheDir, fileName)); err == nil {
+		fmt.Println("Found previous results for ", fileName, b.benchFunc, "Reusing.")
+		return filepath.Join(b.resultCacheDir, fileName), nil
+	}
+
+	// TODO(bwplotka): Allow memprofiles.
+	extraArgs := []string{"-benchtime", b.benchTime.String()}
+	extraArgs = append(extraArgs, "-timeout", b.benchTimeout.String())
+	benchCmd := []string{"bash", "-c", strings.Join(append(append([]string{"cd", pkgRoot, "&&", "go", "test", "-run", "^$", "-bench", fmt.Sprintf("^%s$", b.benchFunc), "-benchmem"}, extraArgs...), "./..."), " ")}
```

Comment:
The real command to be executed is difficult to know, can we add a log here to print the whole command in verbose mode? or do
```go
cmdStr := "cd /some/dir && go test -run . -bench ."
benchCmd := string.Split(cmdStr)
```
probably `func (c *commander) exec` is a better place to log.

Reply:
We have the command logged below